File: python/src/sec_edgar_toolkit/types/institutional_holdings.py
"""Type definitions for 13F institutional holdings parsing."""

from __future__ import annotations

from datetime import datetime
from typing import List, Optional, TypedDict


class SharesOrPrincipalAmount(TypedDict):
    """Shares or principal amount information."""

    shares: float
    principal: float
    type: str  # 'SH', 'PRN'


class VotingAuthority(TypedDict):
    """Voting authority information."""

    sole: int
    shared: int
    none: int


class Holding(TypedDict):
    """Individual holding information."""

    name_of_issuer: str
    title_of_class: str
    cusip: str
    value: float
    shares_or_principal_amount: SharesOrPrincipalAmount
    put_call: Optional[str]  # 'Put', 'Call', None
    investment_discretion: str  # 'SOLE', 'SHARED', 'NONE'
    manager_class: str
    voting_authority: VotingAuthority


class PortfolioSummary(TypedDict):
    """Portfolio summary statistics."""

    total_value: float
    total_positions: int
    top_sector_allocation: str
    concentration_ratio: float
    average_position_size: float


class Position(TypedDict):
    """Individual position with portfolio weight."""

    holding: Holding
    portfolio_weight: float
    rank: int


class SectorAllocation(TypedDict):
    """Sector allocation information."""

    sector: str
    value: float
    percentage: float
    number_of_holdings: int


class HoldingComparison(TypedDict):
    """Comparison between two holdings periods."""

    holding: str
    cusip: str
    previous_value: float
    current_value: float
    change_in_value: float
    change_in_shares: float
    action: str  # 'NEW', 'SOLD_OUT', 'INCREASED', 'DECREASED', 'NO_CHANGE'


class ParsedInstitutionalHolding(TypedDict):
    """Complete parsed institutional holding data."""

    form_type: str
    filing_date: datetime
    period_of_report: datetime
    cik: str
    manager_name: str
    holdings: List[Holding]
    portfolio_summary: PortfolioSummary
    sector_allocations: List[SectorAllocation]

File: typescript/src/types/institutional-holdings.ts
/**
 * Type definitions for 13F institutional holdings parsing
 */

export interface Holding {
  nameOfIssuer: string;
  titleOfClass: string;
  cusip: string;
  value: number;
  sharesOrPrincipalAmount: {
    shares: number;
    principal: number;
    type: 'SH' | 'PRN';
  };
  putCall: 'Put' | 'Call' | null;
  investmentDiscretion: 'SOLE' | 'SHARED' | 'NONE';
  managerClass: string;
  votingAuthority: {
    sole: number;
    shared: number;
    none: number;
  };
}

export interface PortfolioSummary {
  totalValue: number;
  totalPositions: number;
  topSectorAllocation: string;
  concentrationRatio: number;
  averagePositionSize: number;
}

export interface Position {
  holding: Holding;
  portfolioWeight: number;
  rank: number;
}

export interface SectorAllocation {
  sector: string;
  value: number;
  percentage: number;
  numberOfHoldings: number;
}

export interface HoldingComparison {
  holding: string;
  cusip: string;
  previousValue: number;
  currentValue: number;
  changeInValue: number;
  changeInShares: number;
  action: 'NEW' | 'SOLD_OUT' | 'INCREASED' | 'DECREASED' | 'NO_CHANGE';
}

export interface ParsedInstitutionalHolding {
  formType: string;
  filingDate: Date;
  periodOfReport: Date;
  cik: string;
  managerName: string;
  holdings: Holding[];
  portfolioSummary: PortfolioSummary;
  sectorAllocations: SectorAllocation[];
}